File: GeeksforGeeks_POTD/June_2026/June_24.java
package GeeksforGeeks_POTD.June_2026;

/*
Recursion or dfs with modification to explore first shorter to longer and first right thwn down.
For optimization use a boolean array to know that this can't be included in path(Because we may explore same sub-path from different routes which didn't reach end)

TC - O(n^3){Each state(n^2) resolved once with n iterations}, SC - O(n^2){Memorization + Recursion Stack}
*/

import java.util.stream.Stream;
import java.util.stream.Collectors;

class June_24 {
    private boolean dp[][];
    public ArrayList<ArrayList<Integer>> shortestDist(int[][] mat) {
        int n=mat.length-1;
        if(mat[0][0]==0) return new ArrayList<>(List.of(new ArrayList<>(List.of(-1))));
        ArrayList<ArrayList<Integer>> ans= Stream.generate(() -> new ArrayList<>(Collections.nCopies(n+1, 0))) .limit(n+1) .collect(Collectors.toCollection(ArrayList::new));
        dp=new boolean[n+1][n+1];
        if(dfs(n,mat,ans,0,0)) return ans;
        return new ArrayList<>(List.of(new ArrayList<>(List.of(-1))));
    }
    private boolean dfs(int n, int[][] mat, ArrayList<ArrayList<Integer>> ans, int i, int j){
        if(dp[i][j]) return false;
        if(i+j==2*n) {
            ans.get(n).set(n,1);
            return true;
        }
        for(int k=1;k<=mat[i][j];k++){
            if(j+k<=n && mat[i][j+k]!=0){
                if(dfs(n, mat, ans, i, j+k)){
                    ans.get(i).set(j,1);
                    return true;
                }
                else dp[i][j+k]=true;
            }
            if(i+k<=n && mat[i+k][j]!=0){
                if(dfs(n, mat, ans, i+k, j)){
                    ans.get(i).set(j,1);
                    return true;
                }
                else dp[i+k][j]=true;
            }
            if(i+k>n && j+k>n) {
                dp[i][j]=true;
                return false;
            }
        }
        dp[i][j]=true;
        return false;
    }
}

File: GeeksforGeeks_POTD/June_2026/June_24.py
class Solution:
    dp=[[]]
	def shortestDist(self, mat):
	    n=len(mat)-1
		self.dp=[[False for _ in range(n+1)] for _ in range(n+1)]
        if mat[0][0]==0: return [[-1]]
        ans=[[0 for _ in range(n+1)] for _ in range(n+1)]
        if self.dfs(n,mat,ans,0,0): return ans
        return [[-1]]
    def dfs(self, n, mat, ans, i, j):
        if self.dp[i][j]: return False
        if i+j==2*n:
            ans[n][n]=1
            return True
        for k in range(mat[i][j]):
            if j+k<n and mat[i][j+k+1]!=0:
                if self.dfs(n, mat, ans, i, j+k+1):
                    ans[i][j]=1
                    return True
                else: self.dp[i][j+k+1]=True
            if i+k<n and mat[i+k+1][j]!=0:
                if self.dfs(n, mat, ans, i+k+1, j):
                    ans[i][j]=1
                    return True
                else: self.dp[i+k][j]=True
            if i+k>=n and j+k>=n:
                self.dp[i][j]=True
                return False
        self.dp[i][j]=True
        return False